akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala

```diff
@@ -1723,14 +1750,22 @@ class ShardCoordinator(handOffTimeout: FiniteDuration, shardStartTimeout: Finite
       }
 
     case RebalanceTick ⇒
-      if (persistentState.regions.nonEmpty)
-        allocationStrategy.rebalance(persistentState.regions, rebalanceInProgress).foreach { shard ⇒
-          rebalanceInProgress += shard
-          val rebalanceFromRegion = persistentState.shards(shard)
-          log.debug("Rebalance shard [{}] from [{}]", shard, rebalanceFromRegion)
-          context.actorOf(rebalanceWorkerProps(shard, rebalanceFromRegion, handOffTimeout,
-            persistentState.regions.keySet ++ persistentState.regionProxies))
+      if (persistentState.regions.nonEmpty) {
+        val shardsFuture = allocationStrategy.rebalance(persistentState.regions, rebalanceInProgress)
+        shardsFuture.value match {
+          case Some(Success(shards)) ⇒
+            continueRebalance(shards)
+          case _ ⇒
+            // continue when future is completed
+            shardsFuture.map { shards ⇒ RebalanceResult(shards)
+            }.recover {
+              case _ ⇒ RebalanceResult(Set.empty)
+            }.pipeTo(self)
         }
+      }
+
+    case RebalanceResult(shards) ⇒
+      continueRebalance(shards)
```

Comment:
what happens if there are multiple rebalance Futures in flight? The ordering will not be guaranteed between them, can that be a problem? (I looked at the tests but I am unsure if this scenario is covered or not)